feat(medication): EPI-1155: Add warning text when prescribing a medication that has a current prescription

huytranbes opened the pull request:
### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [ ] **Run E2E Tests** <!-- #e2e -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added warnings to alert users if a medication has already been prescribed in the current encounter, both in the medication form and medication list.
- **Style**
  - Improved spacing between medication list items and adjusted heading margins for better readability.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

huytranbes committed:
feat(medication): Enhance medication components with existingDrugIds prop

- Added existingDrugIds prop to MedicationModal, MedicationSetModal, and EncounterMedicationPane for improved medication management.
- Updated MedicationSetList to include SmallBodyText for better display of medication details.
- Refactored MedicationForm to show a warning for existing drugs, enhancing user awareness during medication entry.

huytranbes committed:
fix(medication): Update MedicationListItem key handling in MedicationSetList component

- Removed redundant key prop from MedicationListItem in MedicationSetList to streamline rendering and improve performance.
- Ensured that the key is only set on the parent div for proper React list rendering.

khoi-bes commented:
I think we should use the useEncounterMedicationQuery hook to get all the active medications of an encounter instead of passing prop though all child components

huytranbes committed:
resolve feedback

huytranbes committed:
Merge branch 'main' into feature/epi-1155-add-warning-text-when-prescribing-a-medication-that-has-a-current

khoi-bes commented on packages/web/app/forms/MedicationForm.jsx:
```diff
@@ -522,6 +524,10 @@ export const MedicationForm = ({
   const queryClient = useQueryClient();
   const { loadEncounter } = useEncounter();
+  const { data: { data: medications = [] } = {} } = useEncounterMedicationQuery(encounterId);
+  const existingDrugIds = useMemo(() => medications.map(({ medication }) => medication?.id), [
```
it isn't necessary for using useMemo here

khoi-bes commented on packages/web/app/forms/MedicationForm.jsx:
```diff
@@ -622,6 +629,11 @@ export const MedicationForm = ({
     };
   };
 
+  const onDrugChange = drugId => {
```
can you pls change this function name to `handleChangeMedication` to make it be similar to this function [here](https://github.com/beyondessential/tamanu/pull/7830/files#diff-443417a005e3d3e733072f222b643a60bd35c7696dd2687a75a77b5a4de4a1fbR630) then we can fix the conflicts after that to make sure the code is consistent?

khoi-bes commented on packages/web/app/components/Medication/MedicationTable.jsx:
```diff
@@ -1,12 +1,12 @@
 import { getDose, getTranslatedFrequency } from '@tamanu/shared/utils/medication';
 
-import { DataFetchingTable } from '../Table';
+import { Table } from '../Table';
```
Are there any reasons for these changes?

khoi-bes commented on packages/web/app/components/Medication/MedicationSetList.jsx:
```diff
@@ -131,9 +133,14 @@ export const MedicationSetMedicationsList = ({
   const { getTranslation, getEnumTranslation } = useTranslation();
+  const { encounter } = useEncounter();
+  const { data: { data: medications = [] } = {} } = useEncounterMedicationQuery(encounter?.id);
+  const existingDrugIds = useMemo(() => medications.map(({ medication }) => medication?.id), [
```
it isn't necessary for using useMemo here

huytranbes committed:
resolve feedback

huytranbes committed:
Merge branch 'main' into feature/epi-1155-add-warning-text-when-prescribing-a-medication-that-has-a-current

huytranbes committed:
fix(medication): Update warning text for existing medication prescriptions

- Revised the fallback message in the MedicationSetMedicationsList component to clarify that the medication has already been prescribed for the current encounter, emphasizing the need for clinical appropriateness.

huytranbes committed:
fix(medication): Filter out discontinued medications from existing drug IDs

- Updated the logic in both MedicationSetMedicationsList and MedicationForm components to exclude discontinued medications when generating the list of existing drug IDs.

huytranbes committed:
fix(medication): Add editable condition to existing drug warning display

- Enhanced the MedicationSetMedicationsList component to conditionally display the warning for existing medications only when the editable state is true, improving user experience during medication management.
